### DISCUSSION TITLE
Tolerance, nominal attribute and propagation among variables

### DISCUSSION THREAD

**Discussion by mpradovelasco** [OpenModelica/OpenModelica#10948] [Q&A] [answered]
Hi, 

I need working with very small quantities (with respect to international system of units). OpenModelica seems to solve fine this issue with the tolerance parameter in the simulation stage. However, I am interested to improve the computational efficiency through the use of proper nominal values, assuming that they allow managing the numerical error through: Error_v = tolerance (abs(v) + nominal(v)).

I may have:
```
SI.MassFlowRate mfv (nominal = 1e-9);
connector inputDrug
  input SI.MassFlowRate mf;  //without tolerance definition
end inputDrug;
inputDrug C;
```
I wonder how the nominal attribute propagates if those variables are related through model equations by mfv = - C.mf, The doubt extends also to other functional relationships among variables. In summary, is there any internal scheme that propagates or define the tolerance in OpenModelica? 

Thank you in advance for help!

**Comment by phannebohm** [comment]
I didn't check but cases like `x = y` or `x = -y` should work already via our alias removal aka RemoveSimpleEquations. There was some discussion in #4515 a while ago on the general case for linear or perhaps even arbitrary expressions but to my knowledge none of this is implemented yet.

**Reply by mpradovelasco** [reply]
Thank you for referring to #4515 . Really, I do not know well the scope of OpenModelica with regard the scaling with nominal values for a good mathematical conditioning. 

In other languages I have prepared the equations in a dimensionless form, assuring this manner the proper range of variables to solve, but that method has disadvantages.  

I have tested in OpenModelica models with mass flow rates as small as 0.001 ug/min, combined with others of 1 - 10 kg/min (different chemicals) with success, but I need to reduce the tolerance value to 1e-12. I have implemented a (more or less) general structure to define their nominal values (with the ability to differentiate them in connected spacial regions), but I have not had success in the task of relaxing the tolerance to the default value (1e-6) in the same scenario. 

For instance, the following equation: 
`mfIN = sum(if pi[j].mf > 0 then pi[j].mf else 0 for j in 1:Ni)+ sum(if po[j].mf > 0 then po[j].mf else 0 for j in 1:No)`, 
requires a very accurate crossover detection to calculate mfIN. The following plot capture shows a good behavior when tolerance is 1e-12:

![image](https://github.com/OpenModelica/OpenModelica/assets/114144098/b0cc544d-b032-47de-970e-3cf0fe51b78a)
![image](https://github.com/OpenModelica/OpenModelica/assets/114144098/6ead9ab8-05b9-4a15-a37a-a666e45563b3)

but when tolerance reduces to 1e-11, it produces a bad behaviour, as shows:

![image](https://github.com/OpenModelica/OpenModelica/assets/114144098/4a247ac9-6cda-4914-80a2-6a41a4c50a05)
![image](https://github.com/OpenModelica/OpenModelica/assets/114144098/c1a8ec03-58ff-4412-9356-8b168e7d2165)

and the things worsen when tolerance increments to 1e-10, 1e-9, etc.

The final situation does not improve by controlling the tolerances of variables mf and mfIN of the equation. I would be grateful if you could refer any document that describes the scope and use of nominal attribute in OpenModelica. 

(Figures were obtained with the latest nightly build OM 1.22 dev).

Thanks again!

**Reply by phannebohm** [reply, marked as answer]
> In other languages I have prepared the equations in a dimensionless form, assuring this manner the proper range of variables to solve, but that method has disadvantages.

Ideally, correct scaling has to be managed by the Modelica tool when good nominal values are provided. I guess we're not fully there yet.

> For instance, the following equation: `mfIN = sum(if pi[j].mf > 0 then pi[j].mf else 0 for j in 1:Ni)+ sum(if po[j].mf > 0 then po[j].mf else 0 for j in 1:No)`, requires a very accurate crossover detection to calculate mfIN.

Yes this is a known issue, see #6970. We are working on that one.

> I would be grateful if you could refer any document that describes the scope and use of nominal attribute in OpenModelica.

Off the top of my head there are three major places where nominal values are important: integration methods, nonlinear solvers, and event detection (which, as I said, has no proper scaling yet). I'm not sure if there is good documentation, but you may search the [OpenModelica Users Guide](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/solving.html).

**Reply by mpradovelasco** [reply]
Thank you very much for all your information @phannebohm. It clarifies this subject. I will be then watching of your advances in #6970 issue.

**Comment by mpradovelasco** [comment]
Hi @phannebohm ,

I was reading your interesting comments about addressing nominal values for sums and differences, and I wonder if the best choice for the nominal value of a sum or difference is simply the maximum nominal value of the two terms, what allows deducing the target variable to compute the crossover.

Considering a = An as and b = Bn bs, where An and Bn are the nominal values of a and b, If they have been properly selected, then the scaled values as and bs are around 1. Then, assuming An > Bn:  

a - b = An (as - Bn/An bs) 

That is, the scaled variable (as - Bn/An bs) is also around 1. This seems the target value to compute the crossover. 

In the case that Bn > An, then:

a - b = Bn (An/Bn as - bs) 

That is, the scaled variable to (An/Bn as - bs) seems the target value to compute the crossover. 

Best regards

**Reply by phannebohm** [reply]
> [...] I wonder if the best choice for the nominal value of a sum or difference is simply the maximum nominal value of the two terms [...]

Yes, that is how I implemented it in #10960. I didn't think of the theoretical underpinning, though. Thanks for the suggestion!
Let's discuss this directly in #6970.
